server/src/internalClusterTest/java/org/opensearch/remotestore/PrimaryTermValidationIT.java

```diff
@@ -0,0 +1,165 @@
+        logger.info("--> network disruption is started");
+        networkDisruption.startDisrupting();
+
+        // Ensure the node which is partitioned is removed from the cluster
+        assertBusy(() -> {
+            NodesInfoResponse response = client(clusterManagerNode).admin().cluster().prepareNodesInfo().get();
+            assertThat(response.getNodes().size(), equalTo(2));
+        });
+
+        // Ensure that the cluster manager has latest information about the index
+        assertBusy(() -> {
+            ClusterHealthResponse clusterHealthResponse = client(clusterManagerNode).admin()
+                .cluster()
+                .health(new ClusterHealthRequest())
+                .actionGet(TimeValue.timeValueSeconds(1));
+            assertTrue(clusterHealthResponse.getIndices().containsKey(INDEX_NAME));
+            ClusterIndexHealth clusterIndexHealth = clusterHealthResponse.getIndices().get(INDEX_NAME);
+            assertEquals(ClusterHealthStatus.YELLOW, clusterHealthResponse.getStatus());
+            assertEquals(1, clusterIndexHealth.getNumberOfShards());
+            assertEquals(1, clusterIndexHealth.getActiveShards());
+            assertEquals(1, clusterIndexHealth.getUnassignedShards());
+            assertEquals(1, clusterIndexHealth.getUnassignedShards());
+            assertEquals(1, clusterIndexHealth.getActivePrimaryShards());
+            assertEquals(ClusterHealthStatus.YELLOW, clusterIndexHealth.getStatus());
+        });
+
+        // Index data to the newly promoted primary
+        indexSameDoc(clusterManagerNode, INDEX_NAME);
+        RefreshResponse refreshResponse = client(clusterManagerNode).admin()
+            .indices()
+            .prepareRefresh(INDEX_NAME)
+            .setIndicesOptions(IndicesOptions.STRICT_EXPAND_OPEN_HIDDEN_FORBID_CLOSED)
+            .execute()
+            .actionGet();
+        assertNoFailures(refreshResponse);
+        assertEquals(1, refreshResponse.getSuccessfulShards());
+        assertHitCount(client(replicaNode).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), numOfDocs + 1);
+
+        // At this point we stop the disruption. Since the follower checker has already failed and cluster manager has removed the node
+        // from cluster, failed node needs to start discovery process by leader checker call. We stop the disruption to allow the failed
+        // node to
+        // communicate with the other node which it assumes has replica.
+        networkDisruption.stopDisrupting();
+
+        // When the index call is made to the stale primary, it makes the primary term validation call to the other node (which
+        // it assumes has the replica node). At this moment, the stale primary realises that it is no more the primary and the caller
+        // received the following exception.
+        ShardNotFoundException exception = assertThrows(ShardNotFoundException.class, () -> indexSameDoc(primaryNode, INDEX_NAME));
```

Comment:
Is it possible to checking indexing doc to `primaryNode` before stopping the n/w disruption?

Reply:
Yes, let me add that.

Reply:
This is what I have observed while adding test for the above case. When we partition off the node from the cluster, and index to the stale primary. It receives the requests, tries to do primary term validation. Since the call fails, it tries to fail the shard by communicating the same to cluster manager. Now, here I am observing that this is infinite loop and no limits on how many maximum retries that can happen. It has already retries 49 times while I am writing this. 
```
[2566-07-03T11:30:14,750][WARN ][o.o.c.c.ClusterFormationFailureHelper] [node_t2] cluster-manager not discovered yet: have discovered [{node_t2}{7YwTbiETR0eCGMp7EwwiPw}{VQ0EvKlaTFafHjK5Ac6qWw}{127.0.0.1}{127.0.0.1:51633}{d}{shard_indexing_pressure_enabled=true}]; discovery will continue using [127.0.0.1:51629] from hosts providers and [{node_t0}{0QjD-IpNQaGCTirPp7sYgQ}{cWH7APaTQHup55lA07TMsg}{127.0.0.1}{127.0.0.1:51629}{m}{shard_indexing_pressure_enabled=true}] from last-known cluster state; node term 1, last-accepted version 10 in term 1
[2566-07-03T11:30:24,754][WARN ][o.o.c.c.ClusterFormationFailureHelper] [node_t2] cluster-manager not discovered yet: have discovered [{node_t2}{7YwTbiETR0eCGMp7EwwiPw}{VQ0EvKlaTFafHjK5Ac6qWw}{127.0.0.1}{127.0.0.1:51633}{d}{shard_indexing_pressure_enabled=true}]; discovery will continue using [127.0.0.1:51629] from hosts providers and [{node_t0}{0QjD-IpNQaGCTirPp7sYgQ}{cWH7APaTQHup55lA07TMsg}{127.0.0.1}{127.0.0.1:51629}{m}{shard_indexing_pressure_enabled=true}] from last-known cluster state; node term 1, last-accepted version 10 in term 1
[2566-07-03T11:30:34,670][WARN ][o.o.c.NodeConnectionsService] [node_t2] failed to connect to {node_t0}{0QjD-IpNQaGCTirPp7sYgQ}{cWH7APaTQHup55lA07TMsg}{127.0.0.1}{127.0.0.1:51629}{m}{shard_indexing_pressure_enabled=true} (tried [49] times)
org.opensearch.transport.ConnectTransportException: [node_t0][127.0.0.1:51629] DISCONNECT: simulated
	at org.opensearch.test.transport.MockTransportService.lambda$addFailToSendNoConnectRule$2(MockTransportService.java:326) ~[classes/:?]
	at org.opensearch.test.transport.StubbableTransport.openConnection(StubbableTransport.java:175) ~[classes/:?]
	at org.opensearch.transport.ClusterConnectionManager.internalOpenConnection(ClusterConnectionManager.java:274) ~[classes/:?]
	at org.opensearch.transport.ClusterConnectionManager.connectToNode(ClusterConnectionManager.java:157) ~[classes/:?]
	at org.opensearch.test.transport.StubbableConnectionManager.connectToNode(StubbableConnectionManager.java:118) ~[classes/:?]
	at org.opensearch.transport.TransportService.connectToNode(TransportService.java:488) ~[classes/:?]
	at org.opensearch.transport.TransportService.connectToNode(TransportService.java:468) ~[classes/:?]
	at org.opensearch.cluster.NodeConnectionsService$ConnectionTarget$1.doRun(NodeConnectionsService.java:344) ~[classes/:?]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[classes/:?]
	at java.util.ArrayList.forEach(ArrayList.java:1511) [?:?]
	at org.opensearch.cluster.NodeConnectionsService.connectDisconnectedTargets(NodeConnectionsService.java:229) [classes/:?]
	at org.opensearch.cluster.NodeConnectionsService$ConnectionChecker.doRun(NodeConnectionsService.java:240) [classes/:?]
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:854) [classes/:?]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:833) [?:?]
[2566-07-03T11:30:34,760][WARN ][o.o.c.c.ClusterFormationFailureHelper] [node_t2] cluster-manager not discovered yet: have discovered [{node_t2}{7YwTbiETR0eCGMp7EwwiPw}{VQ0EvKlaTFafHjK5Ac6qWw}{127.0.0.1}{127.0.0.1:51633}{d}{shard_indexing_pressure_enabled=true}]; discovery will continue using [127.0.0.1:51629] from hosts providers and [{node_t0}{0QjD-IpNQaGCTirPp7sYgQ}{cWH7APaTQHup55lA07TMsg}{127.0.0.1}{127.0.0.1:51629}{m}{shard_indexing_pressure_enabled=true}] from last-known cluster state; node term 1, last-accepted version 10 in term 1
[2566-07-03T11:30:44,764][WARN ][o.o.c.c.ClusterFormationFailureHelper] [node_t2] cluster-manager not discovered yet: have discovered [{node_t2}{7YwTbiETR0eCGMp7EwwiPw}{VQ0EvKlaTFafHjK5Ac6qWw}{127.0.0.1}{127.0.0.1:51633}{d}{shard_indexing_pressure_enabled=true}]; discovery will continue using [127.0.0.1:51629] from hosts providers and [{node_t0}{0QjD-IpNQaGCTirPp7sYgQ}{cWH7APaTQHup55lA07TMsg}{127.0.0.1}{127.0.0.1:51629}{m}{shard_indexing_pressure_enabled=true}] from last-known cluster state; node term 1, last-accepted version 10 in term 1
[2566-07-03T11:30:54,770][WARN ][o.o.c.c.ClusterFormationFailureHelper] [node_t2] cluster-manager not discovered yet: have discovered [{node_t2}{7YwTbiETR0eCGMp7EwwiPw}{VQ0EvKlaTFafHjK5Ac6qWw}{127.0.0.1}{127.0.0.1:51633}{d}{shard_indexing_pressure_enabled=true}]; discovery will continue using [127.0.0.1:51629] from hosts providers and [{node_t0}{0QjD-IpNQaGCTirPp7sYgQ}{cWH7APaTQHup55lA07TMsg}{127.0.0.1}{127.0.0.1:51629}{m}{shard_indexing_pressure_enabled=true}] from last-known cluster state; node term 1, last-accepted version 10 in term 1
[2566-07-03T11:31:04,774][WARN ][o.o.c.c.ClusterFormationFailureHelper] [node_t2] cluster-manager not discovered yet: have discovered [{node_t2}{7YwTbiETR0eCGMp7EwwiPw}{VQ0EvKlaTFafHjK5Ac6qWw}{127.0.0.1}{127.0.0.1:51633}{d}{shard_indexing_pressure_enabled=true}]; discovery will continue using [127.0.0.1:51629] from hosts providers and [{node_t0}{0QjD-IpNQaGCTirPp7sYgQ}{cWH7APaTQHup55lA07TMsg}{127.0.0.1}{127.0.0.1:51629}{m}{shard_indexing_pressure_enabled=true}] from last-known cluster state; node term 1, last-accepted version 10 in term 1
```
From the caller's perspective, there will be no response until the stale primary is able to successfully communicate with the cluster manager node for failing the replica shard (from it's own perspective). Let me know if you have any ideas around how to test this. Btw this is a status quo behavior today in replication.

Reply:
So, does this mean the call will timeout at caller's end?